### ISSUE TITLE
Reasoning behind constraint in `grads_and_grad_moms`

### ISSUE THREAD

**Issue by aomader** [ProbabilisticNumerics/probabilistic_line_search#1]
Hej,

I'm wondering about the reasoning behind [this constraint](https://github.com/ProbabilisticNumerics/probabilistic_line_search/blob/master/probls/tensorflow_interface/gradient_moment.py#L89-L91) in `grads_and_grad_moms`.
The given variables are indeed used in multiple operations in the loss computation graph, nonetheless that shouldn't hinder it to compute the grad.

Could you shed some light on this?

**Comment by lballes** [comment]
Hey!

The second moment of gradients, which is needed for the gradient variance computation, is computed implicitly using a "trick" explained in this [note][1]. This trick is not directly applicable when a variable has more than one incoming gradients during back-propagation, which is the case whenever a variable is consumed by more than one operation.

Since we currently don't have an efficient way to compute the gradient moment in such a case, we ruled out multiple consumer operations. We are trying to come up with a better implementation of the gradient moment computation; any ideas/help would be highly appreciated :)

[1]: https://drive.google.com/file/d/0B0adgqwcMJK5aDNaQ2Q4ZmhCQzA/view

**Comment by aomader** [comment]
@lballes I would say the most straight forward way would be to fall back to an inefficient method to compute the second moment, but `warnings.warn()` the user that the trick doesn't apply to the supplied objective function.

**Comment by hughsalimbeni** [comment]
Hi, have you tried a running average estimate like in ADAM:
```
m_t <- beta1 * m_{t-1} + (1 - beta1) * g
v_t <- beta2 * v_{t-1} + (1 - beta2) * g * g
```
(taken [from here](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer))

The requirement for all operations to be `matmul`, `add` and `conv2d` seems rather restrictive

**Comment by lballes** [comment]
Hi @hughsalimbeni, we did some general experiments comparing running average estimates to mini-batch estimates and found the former to be quite poor. That's why we went with the mini-batch estimate for this implementation, which is of course very restrictive for practical use in its current implementation. It would definitely be worth exploring how the line search would actually perform using the running average estimates. I don't know whether I personally will find time to do that anytime soon, though.
